packages/arb-bridge-eth/contracts/rollup/facets/RollupAdmin.sol

```diff
@@ -269,53 +274,77 @@ contract RollupAdminFacet is RollupBase, IRollupAdmin {
 
         RollupLib.Assertion memory assertion =
-                RollupLib.decodeAssertion(
-                    assertionBytes32Fields,
-                    assertionIntFields,
-                    beforeProposedBlock,
-                    beforeInboxMaxCount,
-                    sequencerBridge.messageCount()
-                );
-
-        bytes32 nodeHash =
-            _newNode(
+            RollupLib.decodeAssertion(
+                assertionBytes32Fields,
+                assertionIntFields,
+                beforeProposedBlock,
+                beforeInboxMaxCount,
+                sequencerBridge.messageCount()
+            );
+
+        uint256 deadlineBlock =
+            max(block.number.add(confirmPeriodBlocks), getNode(prevNode).deadlineBlock());
+
+        (bytes32 nodeHash, ) =
+            createNewNode(
                 assertion,
                 deadlineBlock,
                 sequencerBatchEnd,
                 sequencerBatchAcc,
                 prevNode,
                 getNodeHash(prevNode),
-                false
+                false,
+                NewNodeDependencies({
+                    sequencerInbox: sequencerBridge,
+                    rollupEventBridge: rollupEventBridge,
+                    nodeFactory: nodeFactory
+                })
             );
-        // TODO: should we add a stake?
-        
+
         require(expectedNodeHash == nodeHash, "NOT_EXPECTED_HASH");
+
+        stakeOnNode(msg.sender, latestNodeCreated(), confirmPeriodBlocks);
+
+        emit OwnerFunctionCalled(23);
     }
 
     function forceConfirmNode(
+        uint256 nodeNum,
+        bytes32 beforeSendAcc,
         bytes calldata sendsData,
-        uint256[] calldata sendLengths
+        uint256[] calldata sendLengths,
+        uint256 afterSendCount,
+        bytes32 afterLogAcc,
+        uint256 afterLogCount
     ) external override whenPaused {
+        bytes32 afterSendAcc = RollupLib.feedAccumulator(sendsData, sendLengths, beforeSendAcc);
+
+        INode node = getNode(nodeNum);
+
+        require(
+            node.confirmData() ==
+                RollupLib.confirmHash(
+                    beforeSendAcc,
+                    afterSendAcc,
+                    afterLogAcc,
+                    afterSendCount,
+                    afterLogCount
+                ),
+            "CONFIRM_DATA"
+        );
+        // processes outgoing messages without node.requirePastDeadline();
         outbox.processOutgoingMessages(sendsData, sendLengths);
 
-        confirmLatestNode();
-
-        rollupEventBridge.nodeConfirmed(latestConfirmed());
+        confirmNode(nodeNum);
+        rollupEventBridge.nodeConfirmed(nodeNum);
 
-        // emit NodeConfirmed(
-        //     firstUnresolved,
-        //     afterSendAcc,
-        //     afterSendCount,
-        //     afterLogAcc,
-        //     afterLogCount
-        // );
+        emit NodeConfirmed(nodeNum, afterSendAcc, afterSendCount, afterLogAcc, afterLogCount);
```

Comment:
If possible it'd be nice to do some refactoring to reduce the duplication between the regular rollup methods and the admin methods

Reply:
Refactored for the `create new node` and `confirm node` to use more shared logic in the RollupCore.

The non-shared code is for the checks that the user rollup does that are skipped in the admin - should be easy to throw them around now.